Cleanup and reorganization of `nbt`.

atheriel opened the pull request:
Originally, I wanted to clean up some stray `panic!`'s and add some more tests (in line with my comments on IRC), but in the process I tried simplifying and improving the whole module. I though I might submit it for review. Let me know what you think.

Generally speaking, I removed quite a lot of redundant code and rewrote & simplified the readers and writers to make them recursive. There's now a new top-level object, `NbtFile` (which could also be renamed, say, `NbtObject`) which encapsulates the idea of all NBT-encoded data being a named `Compound` and handles the high-level encoding/decoding/indexing.

There are also some new and more extensive tests. I've used them to verify that the new parsers match the output of the old ones.

Finally, the use of the `flate` crate has been replaced by the newer, better `flate2` one.

atheriel committed:
Removes panic! calls from the Nbt implementation.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
Adds tests for some Nbt decoding errors.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
Implements a new (cleaner, much simpler) `nbt` API.

Major changes:

* Consolidates the `Nbt` and `List` objects into one enum, `NbtValue`.
* Reading and writing of `NbtValue`s has now been consolidated into two
  recursion-friendly methods, instead of the old private `read_i8`, etc
  approach.
* Introduces a new high-level object, `NbtFile`, which encapsulates a
  named `Compound`, and should be the object to work with in most cases
  instead of the `NbtValue`. Compressed file readers have been moved to
  this object, for example.
* `NbtFile` now implements `Index` and `Protocol`, and the logic inside
  of these implementations has been moved into `NbtFile` itself.
* `Nbt::End` has been abstracted away. It only appears in the reading
  and writing methods now. This should encourage correctness.

Tests are currently very broken. Documentation is also forthcoming.

Hopefully Rust will get updated soon and the `(&mut *x)` borrow crud
can be removed.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
Corrects the tests in the `nbt` module.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
Adds tests for `Nbt*` objects correctly reporting their length.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
Adds additional tests of encoding for `Nbt*` objects.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

toqueteos commented:
Really awesome patch @atheriel!

You can get rid of ~~most~~ all of those `(&mut *x)` by making `Read` and `Write` sinks `mut` as in:

``` rust
pub fn write_header(&self, mut sink: &mut io::Write, title: &String) -> io::Result<()> {
                           ^^^
```

Here's nbt.rs working without any (&mut *x) https://gist.github.com/toqueteos/7c737b54293a2a595656

**EDIT:** Why does GH say we can't auto-merge this?

fenhl commented on src/types/nbt.rs:
```diff
@@ -25,377 +25,291 @@ pub enum Nbt {
-        }
-        Ok(())
+    fn proto_encode(value: &NbtFile, dst: &mut io::Write) -> io::Result<()> {
+        value.write(&mut *dst)
```
Why not just `value.write(dst)`?

atheriel replied:
It gives an `Error: cannot borrow ... as mutable` without it. @toqueteos pointed out another way to fix it, which I'll use instead.

fenhl commented on src/types/nbt.rs:
```diff
@@ -25,377 +25,291 @@ pub enum Nbt {
-            value => panic!("Unknown NBT tag value {}", value)
-        }
+    fn proto_decode(src: &mut io::Read) -> io::Result<NbtFile> {
+        NbtFile::from_reader(&mut *src)
```
Why not just `NbtFile::from_reader(src)`?

fenhl commented:
While removing the `List` type makes the code cleaner, it also no longer enforces the invariant that all elements of a list must be of the same type. I'm not sure if that's a good idea.

atheriel commented:
Yes, it is a tradeoff, and something of a limitation of Rust's enum variants, imo. The other way of enforcing it is during reading/writing and when using the `.insert()` method on `NbtFile`, similar to how all NBT root nodes being Compounds is done. This would ensure you could never read/write/create an incorrect lists, even though the type system in theory allows you to. I thought this would be more consistent.

I'll add those checks in a moment so you can see what they look like.

atheriel commented:
Rebased onto master to fix the merge problem.

atheriel committed:
nbt: Removes the `&mut *` borrows.

atheriel committed:
nbt: Forbid heterogenous List values during creation/writing.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel commented:
@fenhl I've added the checks for lists to prevent incorrect ones.
@toqueteos All the nasty `&mut *` has been excised. Thanks for the tip on `mut`.

toqueteos commented:
The List thingy is a bit worrisome, what if in a near future, when hematite is successful in the modding community, a malevolent uses uses custom Nbt packets to crash entire servers?

atheriel commented:
If the Nbt packet is incorrect in any way, it'll just raise an `io::Errror` when decoding it. Malevolent packets can't crash the server using this code, because I removed all of the calls to `panic!`.

toqueteos commented:
@atheriel So we are forced to check if every item in every List is the same type?

atheriel commented:
Yes. This is actually the same behaviour as before, though. It's just a regular `InvalidInput` error.

The removal of `List` only impedes the ability to use Rust's type system to prevent creating incorrect packets at compile time via instantiation of `NbtValue` variables.

atheriel commented:
There should only be a performance penalty during calls to `write()` and `insert()`.

atheriel committed:
nbt: Now uses flate2 to wrap `io::Read` and `io::Write` when needed.

This also expands writing methods to allow writing compressed `NbtFile`
objects, and includes a test to check that inflation/deflation matches.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
nbt: Removes some stray `&mut *`.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

fenhl commented on src/types/nbt.rs:
```diff
@@ -25,377 +27,309 @@ pub enum Nbt {
-        Ok(v)
-    }
+#[derive(Clone, Debug, PartialEq)]
+pub struct NbtFile {
```
As has been suggested on IRC, we might want to rename `NbtFile` to `NbtBlob`.

atheriel committed:
nbt: Adds documentation for public `Nbt*` methods.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
nbt: Renames `NbtFile` to `NbtBlob`.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel commented:
@fenhl renaming complete.
